marlowe/src/Language/Marlowe/Common.hs

```diff
@@ -0,0 +1,845 @@
+            | TrueObs
+            | FalseObs
+            deriving (Eq, Show)
+
+{-| Marlowe Contract Data Type
+-}
+data Contract = Null
+            | CommitCash IdentCC PubKey Value Timeout Timeout Contract Contract
+            -- ^ commit identifier, owner, amount,
+            --   start timeout, end timeout, OK contract, timed out contract
+            | RedeemCC IdentCC Contract
+            -- ^ commit identifier to redeem, continuation contract
+            | Pay IdentPay Person Person Value Timeout Contract
+            -- ^ pay identifier, from, to, amount, payment timeout,
+            --   continuation (either timeout or successful payment)
+            | Both Contract Contract
+            -- ^ evaluate both contracts
+            | Choice Observation Contract Contract
+            -- ^ if observation evaluates to True evaluates first contract, second otherwise
+            | When Observation Timeout Contract Contract
+            -- ^ when observation evaluates to True evaluate first contract,
+            --   evaluate second contract on timeout
+            deriving (Eq, Show)
+
+{-|
+    State of a contract validation function.
+-}
+data ValidatorState = ValidatorState {
+        maxCCId  :: Int,
+        maxPayId :: Int
+    }
+
+{-|
+    Internal Marlowe Contract state.
+    Persisted in Data Script.
+-}
+data State = State {
+        stateCommitted  :: [Commit],
+        stateChoices :: [Choice]
+    } deriving (Eq, Ord, Show)
+
+emptyState :: State
+emptyState = State { stateCommitted = [], stateChoices = [] }
+
+{-|
+    Contract input command.
+    'Commit', 'Payment', and 'Redeem' all require a proof
+    that the transaction is issued by a particular party identified with /Public Key/.
+    We require 'Signature' of TxHash signed with that /Public Key/.
+
+    E.g. if we have
+    @ CommitCash ident pubKey (Value 100) ... @
+    then we require
+    @ Commit ident signature(pubKey) @
+    to validate that transaction.
+-}
+data InputCommand = Commit IdentCC Signature
+    | Payment IdentPay Signature
+    | Redeem IdentCC Signature
+    | SpendDeposit Signature
+makeLift ''InputCommand
+
+{-|
+    Marlowe Contract Input.
+    May contain oracle values, and newly made choices.
+-}
+data Input = Input InputCommand [OracleValue Int] [Choice]
+
+{-|
+    This data type is a content of a contract's /Data Script/
+-}
+data MarloweData = MarloweData {
+        marloweState :: State,
+        marloweContract :: Contract
+    }
+
+makeLift ''IdentCC
+makeLift ''IdentChoice
+makeLift ''IdentPay
+makeLift ''CCRedeemStatus
+makeLift ''Value
+makeLift ''Observation
+makeLift ''Contract
+makeLift ''ValidatorState
+makeLift ''MarloweData
+makeLift ''Input
+makeLift ''State
+
+-- | 'IdentCC' equality
+eqIdentCC :: Q (TExp (IdentCC -> IdentCC -> Bool))
+eqIdentCC = [|| \(IdentCC a) (IdentCC b) -> a == b ||]
+
+-- | 'Value' equality
+equalValue :: Q (TExp (Value -> Value -> Bool))
```

Comment:
I'm sorry you can't use `deriving Eq`, I think this indeed what you have to do :(

Reply:
No problem, we are not expecting you to add support for all the Haskell's goodies :)
Although, that one would be neat. Can we special-case for `Eq` typeclass?

Reply:
The problem is that we generally don't have access to the definitions of functions. So we can't see the definition of `==`, and if we could we couldn't see the definitions of the functions that were used to define it (which there would be many of, if it was inlined).

Reply:
Fair enough. I had similar problems in JScala